lib/ansible/modules/cloud/azure/azure_rm_virtualmachine_facts.py

```diff
@@ -0,0 +1,228 @@
+#!/usr/bin/python
+#
+# Copyright (c) 2018
+# Gustavo Muniz do Carmo <gustavo@esign.com.br>
+# Zim Kalinowski <zikalino@microsoft.com>
+#
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+
+DOCUMENTATION = '''
+---
+module: azure_rm_virtualmachine_facts
+
+version_added: "2.7"
+
+short_description: Get virtual machine facts.
+
+description:
+  - Get facts for all virtual machines of a resource group.
+
+options:
+    resource_group:
+        description:
+        - Name of the resource group containing the virtual machines (required when filtering by vm name).
+    name:
+        description:
+        - Name of the virtual machine.
+    tags:
+        description:
+        - Limit results by providing a list of tags. Format tags as 'key' or 'key:value'.
+
+extends_documentation_fragment:
+  - azure
+
+author:
+  - "Gustavo Muniz do Carmo (@gustavomcarmo)"
+  - "Zim Kalinowski (@zikalino)"
+
+'''
+
+EXAMPLES = '''
+  - name: Get facts for all virtual machines of a resource group
+    azure_rm_virtualmachine_facts:
+      resource_group: Testing
+
+  - name: Get facts by name
+    azure_rm_virtualmachine_facts:
+      resource_group: Testing
+      name: vm
+
+  - name: Get facts by tags
+    azure_rm_virtualmachine_facts:
+      resource_group: Testing
+      tags:
+        - testing
+        - foo:bar
+'''
+
+RETURN = '''
+azure_virtualmachines:
+    description: List of resource group's virtual machines dicts.
+    returned: always
+    type: list
+    example: [{}]
+'''
+
+try:
+    from msrestazure.azure_exceptions import CloudError
+except:
+    # This is handled in azure_rm_common
+    pass
+
+from ansible.module_utils.azure_rm_common import AzureRMModuleBase, azure_id_to_dict
+from ansible.module_utils.common.dict_transformations import camel_dict_to_snake_dict
+from ansible.module_utils.six.moves.urllib.parse import urlparse
+import re
+
+
+AZURE_OBJECT_CLASS = 'VirtualMachine'
+
+AZURE_ENUM_MODULES = ['azure.mgmt.compute.models']
+
+
+class AzureRMVirtualMachineFacts(AzureRMModuleBase):
+
+    def __init__(self):
+
+        self.module_arg_spec = dict(
+            resource_group=dict(type='str'),
+            name=dict(type='str'),
+            tags=dict(type='list')
+        )
+
+        self.results = dict(
+            changed=False,
+            ansible_facts=dict(azure_virtualmachines=[])
```

Comment:
Don't return the value as part of `ansible_facts`, this should be reserved for the remote host (executing the module). We are getting facts on another resource that's not necessarily related to the executing host.

Because these aren't stored as part of the facts we can shorten the name of the key. This is because there is no need to let them know this is Azure VMs, they already know because they ran an Azure module and registered the result. I would personally use `vms` as the key but up to you if you want to use another name.

tldr: just do

```
self.results = dict(
    changed=False,
    vms=[]
```